sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEOptimizer.scala

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.RuleExecutor
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.util.Utils
+
+/**
+ * The optimizer for re-optimizing the logical plan used by AdaptiveSparkPlanExec.
+ */
+class AQEOptimizer(conf: SQLConf) extends RuleExecutor[LogicalPlan] {
+  private val defaultBatches = Seq(
+    Batch("Demote BroadcastHashJoin", Once,
+      DemoteBroadcastHashJoin(conf)),
+    Batch("Eliminate Join to Empty Relation", Once, EliminateJoinToEmptyRelation)
+  )
+
+  override protected def batches: Seq[Batch] = {
```

Comment:
Should we mark it as a `final` to be consistent with `Optimizer`?

Reply:
sounds good